optimizing constant expressions in profile parsing

abrensch opened the pull request:
This is the patch for the corresponding issue https://github.com/abrensch/brouter/issues/566

It improves the handling of constant expressions in the profile compiler

Visible effect should be that tags that are unused by configuration (e.g. estimated_forest_class) do not appear in the table if process_unused_tags=false

Tested so far via the unit tests only.

Unit tests had found a problem with injected key-values so I had to change the handling of these. An Injected value is now injected as a Number-Expression-Operand of the assign-expression that assigns the variable.

afischerdev commented on brouter-core/src/main/java/btools/router/RoutingContext.java:
```diff
@@ -108,16 +108,6 @@ public long getKeyValueChecksum() {
 
   public void readGlobalConfig() {
     BExpressionContext expctxGlobal = expctxWay; // just one of them...
-
-    if (keyValues != null) {
-      // add parameter to context
-      for (Map.Entry<String, String> e : keyValues.entrySet()) {
-        float f = Float.parseFloat(e.getValue());
-        expctxWay.setVariableValue(e.getKey(), f, true);
-        expctxNode.setVariableValue(e.getKey(), f, true);
-      }
-    }
-
```
Please keep this part to add extra parameter.

afischerdev commented on brouter-expressions/src/main/java/btools/expressions/BExpressionContext.java:
```diff
@@ -857,6 +866,7 @@ int getVariableIdx(String name, boolean create) {
         num = variableNumbers.size();
         variableNumbers.put(name, num);
+        lastAssignedExpression.add(null);
       } else {
```
When you keep the parameter inserts, this needs a Null check like
if (lastAssignedExpression != null) lastAssignedExpression.add(null);

abrensch commented on brouter-core/src/main/java/btools/router/RoutingContext.java:
Got it. I reworked the keyValue injection, but in a way that keeps a single data-path and works at parsetime only.

I also added a unit-test, and extended the "ProfileComparator" to support manual testing.

abrensch commented:
Repeating my last comment that is hidden in the "resolved conversation":

abrensch commented:
Repeating my last comment that is hidden in the "resolved conversation": 

Got [your point about injecting keys not defined in the profile]. I reworked the keyValue injection, but in a way that keeps a single data-path and works at parsetime only.

I also added a unit-test, and extended the "ProfileComparator" to support manual testing.

afischerdev commented:
Great, this worked again for me and my tests.
Only 'consider_traffic' in fastbike profile should be reworked. This is not reachable.
